### PR TITLE
ENH: Return dictionary of contrast images with "all" argument in second-level model

### PR DESCRIPTION
**Pull request by effigies**
Follow-up to #183, doing the same for SecondLevelModel as I did previously for FirstLevelModel.

### CODE REVIEW

**Comment by bthirion** [comment]
... But please fix the test failure...

**Comment by effigies** [comment]
I'll look at this soon, but just posting the traceback for quick reference:

```
Traceback (most recent call last):
  File "/home/travis/miniconda3/envs/testenv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/miniconda3/envs/testenv/lib/python2.7/site-packages/nistats-0.0.1b0-py2.7.egg/nistats/tests/test_second_level_model.py", line 196, in test_second_level_model_contrast_computation
    assert_array_equal(all_images['p_value'].get_fdata(), p_image.get_fdata())
  File "/home/travis/miniconda3/envs/testenv/lib/python2.7/site-packages/numpy/testing/utils.py", line 807, in assert_array_equal
    verbose=verbose, header='Arrays are not equal')
  File "/home/travis/miniconda3/envs/testenv/lib/python2.7/site-packages/numpy/testing/utils.py", line 688, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not equal
(shapes (7, 8, 9, 1), (7, 8, 9) mismatch)
 x: array([[[[  0.000000e+00],
         [  0.000000e+00],
         [  0.000000e+00],...
 y: array([[[  0.000000e+00,   0.000000e+00,   0.000000e+00,   1.000000e+00,
           0.000000e+00,   4.868803e-75,   1.175536e-75,   0.000000e+00,
           0.000000e+00],...
>>  raise AssertionError('\nArrays are not equal\n\n(shapes (7, 8, 9, 1), (7, 8, 9) mismatch)\n x: array([[[[  0.000000e+00],\n         [  0.000000e+00],\n         [  0.000000e+00],...\n y: array([[[  0.000000e+00,   0.000000e+00,   0.000000e+00,   1.000000e+00,\n           0.000000e+00,   4.868803e-75,   1.175536e-75,   0.000000e+00,\n           0.000000e+00],...')
```

**Comment by bthirion** [comment]
@effigies let me know if you can work on it soonish. We're having a nistats sprint atm and I'd be happy to merge that one.

**Comment by effigies** [comment]
Merged master on the chance that the test failure was a one-off or separately fixed. I haven't had a chance to look into why the data shapes of the resulting images would change based on whether they were returned individually or in a dictionary.

File: nistats/second_level_model.py [hunk 1]
```diff
@@ -436,12 +433,17 @@ def compute_contrast(
         contrast = mem_contrast(self.labels_, self.results_, con_val,
                                 second_level_stat_type)
 
-        # We get desired output from contrast object
-        estimate_ = getattr(contrast, output_type)()
+        output_types = valid_types[:-1] if output_type == 'all' else [output_type]
+
+        outputs = {}
+        for output_type_ in output_types:
+            # We get desired output from contrast object
```
Comment (effigies) [inline nistats/second_level_model.py]: @bthirion By inserting `assert False` here, I've managed to show the following effect in PDB:

```Python
ipdb> contrast.p_value().shape
(256,)
ipdb> contrast.stat().shape                                                                         
(256,)
ipdb> contrast.p_value().shape                                                                      
(1, 256)
```

This is reliable. If y'all have time to look at what the `contrast` object is doing where the `stat` method has a side-effect of changing the shapes of the return values of the `p_value` and `z_score` methods, I think that's the heart of the problem. I can try digging more later.

File: nistats/contrasts.py [hunk 2]
```diff
@@ -192,8 +192,8 @@ def stat(self, baseline=0.0):
                 np.maximum(self.variance, self.tiny))
         else:
             raise ValueError('Unknown statistic type')
-        self.stat_ = stat
-        return stat.ravel()
+        self.stat_ = stat.ravel()
+        return self.stat_
```
Comment (effigies) [inline nistats/contrasts.py]: Note that in line 213, `p_value` sets `self.stat_` to the output of `self.stat()`, where here we were outputing an `ravel`'d version of `self.stat_`. This seems to be the heart of the bug.

Comment (bthirion) [reply nistats/contrasts.py]: Indeed ! 
OK, I can try and fix that one.

Comment (effigies) [reply nistats/contrasts.py]: I think this is the fix...

**Comment by bthirion** [comment]
Can you rebase on current master ? normally this should fix travis. Thx

**Comment by effigies** [comment]
Rebased and tests passing.

**Comment by bthirion** [comment]
Great. Mergin, thx.

**Pull request merged**